### PR TITLE
[Model Averaging] Create a post-localSGD communication hook

### PR DESCRIPTION
**Pull request by wayi1**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61207
* **#61206**

Create a communication hook to run post-local SGD. This will be combined with model averager component to better support local SGD.

In contrast to the previous approach that runs local gradient averaging + global model averaging at each step for the first K steps, now we plan to runn global gradient averaging only for the first K steps at each step, just like normal DDP. This can give us two advantages:
1) For some optimizers, model averaging can cause discrepancy in optimizer states. If we still do global gradient averaging for the first K steps, we can defer such discrepancy until we actually start local SGD.
2) Gradient averaging at the first K steps only run one allreduce that overlaps with backward pass, so it should also be more efficient.

Proposal: https://github.com/pytorch/pytorch/issues/59699

Differential Revision: [D29523292](https://our.internmc.facebook.com/intern/diff/D29523292/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29523292/)!

### CODE REVIEW

File: torch/distributed/algorithms/ddp_comm_hooks/post_localSGD_hook.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+    # Run allreduce using `global_group_to_use` in the first `start_localSGD_iter` iterations.
+    if state.iter < state.start_localSGD_iter:
+        state.maybe_increase_iter(bucket)
+        return default._allreduce_fut(global_group_to_use, input_tensor)
+
+    # Run allreduce using `subgroup` after the first `start_localSGD_iter` iterations.
+    # From this moment, model averaging should run after the optimizer step,
+    # to globally allreduce all the parameters.
+    return default._allreduce_fut(state.subgroup, input_tensor)
```
Comment (rohan-varma) [inline torch/distributed/algorithms/ddp_comm_hooks/post_localSGD_hook.py:90]: This hook won't actually result in local SGD being automatically enabled right, the user would still need to manually enable local SGD? It looks the hook controls within which subgroup to run the allreduce. Should we decouple it from local SGD concepts?

Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/post_localSGD_hook.py:90]: > This hook won't actually result in local SGD being automatically enabled right, the user would still need to manually enable local SGD?

I will consider allreduce within subgroups as a kind of local update, which actually is a part of local SGD. Currently the user does need to enable the other model averaging module separately. Once we complete incorporating optimizer into DDP comm hook, then we can further append the model averaging component in the same hook.

> It looks the hook controls within which subgroup to run the allreduce. Should we decouple it from local SGD concepts?

The implementation here actually does not look as generic as it is. Once we don't do global allreduce in the comm hook, we are already doing local update as part of local SGD, and there must be a global model averaging at some time. Therefore, I don't think we need to decouple it from local SGD concept, and I cannot imagine how this hook can be used for other purposes.

File: torch/testing/_internal/distributed/distributed_test.py [hunk 2]
```diff
@@ -3893,6 +3894,23 @@ def test_ddp_hook_parity_powerSGD(self):
...
+        def test_ddp_hook_parity_post_localSGD(self):
+            # Although we start run local SGD at iteration 10, since we still use the global process group to run it,
+            # the post-LocalSGD actually still allreduces gradients globally for the remaining iterations.
+            state = post_localSGD.PostLocalSGDState(process_group=None, subgroup=dist.group.WORLD, start_localSGD_iter=10)
+            self._test_ddp_hook_parity(state=state, hook=post_localSGD.post_localSGD_hook)
+
+            # Since we start local SGD later than the total number of 100 iterations,
+            # no local SGD actually is executed, and we don't even need to provide a subgroup for this case.
+            state = post_localSGD.PostLocalSGDState(process_group=None, subgroup=None, start_localSGD_iter=1000)
+            self._test_ddp_hook_parity(state=state, hook=post_localSGD.post_localSGD_hook)
```
Comment (pritamdamania87) [inline torch/testing/_internal/distributed/distributed_test.py:3912]: Should we also test a scenario where we have different global and subprocess groups? Or is that going to be tested in a follow up PR?

File: torch/distributed/algorithms/ddp_comm_hooks/post_localSGD_hook.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    )
+    world_size = global_group_to_use.size()
+
+    # The input tensor is a flattened 1D tensor.
+    input_tensor = bucket.get_tensor()
+
+    # Run allreduce using `global_group_to_use` in the first `start_localSGD_iter` iterations.
+    if state.iter < state.start_localSGD_iter:
+        state.maybe_increase_iter(bucket)
+        return default._allreduce_fut(global_group_to_use, input_tensor)
```
[the remark this answers is not in the document]
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/post_localSGD_hook.py:85]: In the future we may need to support other local SGD variants, and this method can be easily reused this way.

Personally, since `iter` is a data member of `state`, I prefer the OO flavor here.

File: torch/testing/_internal/distributed/distributed_test.py (hunk 2 above)
Comment (SciPioneer) [reply torch/testing/_internal/distributed/distributed_test.py:3912]: Let me do it in a follow-up PR. To make a difference, I probably need to refactor `_test_ddp_hook_parity` first.

File: torch/distributed/algorithms/ddp_comm_hooks/post_localSGD_hook.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+class PostLocalSGDState(object):
+    r"""
+    Stores the state for all-reducing gradients globally using ``process_group`` until step ``start_localSGD_iter``,
+    and all-reducing gradients locally using ``subgroup`` afterwards.
+    """
+
+    __slots__ = [
+        "process_group",
+        "subgroup",
```
[the remark this answers is not in the document]
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/post_localSGD_hook.py:17]: Explained in L34

File: torch/distributed/algorithms/ddp_comm_hooks/post_localSGD_hook.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+            Note that since DDP comm hook only supports single process single device mode at this time,
+            only exactly one tensor is stored in this bucket.
+
+    Returns:
+        Future handler of the communication, which updates the gradients in place.
+
+    Example::
+        >>> state = PostLocalSGDState(process_group=process_group, subgroup=subgroup,
+                                  start_localSGD_iter=10)
+        >>> ddp_model.register_comm_hook(state, post_localSGD_hook)
```
[the remark this answers is not in the document]
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/post_localSGD_hook.py:72]: What do you think of just adding a link to the docstring of model averaging module, which describes the second half? I don't really want to have duplicate docstring. Otherwise we need to keep them consistent, which can be a burden for maintenance.

File: torch/distributed/algorithms/ddp_comm_hooks/post_localSGD_hook.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pritamdamania87) [reply torch/distributed/algorithms/ddp_comm_hooks/post_localSGD_hook.py:89]: I guess if you want a pure OO flavor here, shouldn't this line `if state.iter < state.start_localSGD_iter:` move into `maybe_increase_iter` or a different method of `state`? I guess what we have now is a mix of OO and something that is not OO.

File: torch/distributed/algorithms/ddp_comm_hooks/post_localSGD_hook.py [hunk 6]
```diff
@@ -70,6 +72,8 @@ def post_localSGD_hook(
         >>> state = PostLocalSGDState(process_group=process_group, subgroup=subgroup,
                                   start_localSGD_iter=10)
         >>> ddp_model.register_comm_hook(state, post_localSGD_hook)
+        >>> # Also need to establish a model averaging module and run model averaging after ``optimizer.step()``.
```
Comment (pritamdamania87) [inline torch/distributed/algorithms/ddp_comm_hooks/post_localSGD_hook.py:75]: Maybe remove the example from here all together and in the docs above just mention that look at the docs of model_averaging for an example. I think that might be more cleaner.

**Review by pritamdamania87 — APPROVED** [review]

File: torch/distributed/algorithms/ddp_comm_hooks/post_localSGD_hook.py (hunk 6 above)
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/post_localSGD_hook.py:75]: Eventually there will be multiple `averager` classes corresponding to different model averaging algorithms, but probably we will only have a single comm hook for the post-localSGD paradigm. Therefore, if we merge this to `PeriodicAverager`, we will also need to merge it to any new `Averager` class.

File: torch/distributed/algorithms/ddp_comm_hooks/post_localSGD_hook.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/post_localSGD_hook.py:89]: If we move `if state.iter < state.start_localSGD_iter:` into a member method, we also need to wrap up `default._allreduce_fut(global_group_to_use, input_tensor)` as a callback. That will cause some unnecessary complexity. I am not a fan of pure OO.

File: torch/testing/_internal/distributed/distributed_test.py [hunk 7]
```diff
@@ -3895,6 +3897,25 @@ def test_ddp_hook_parity_powerSGD(self):
                     state=powersgd_state, hook=powerSGD.powerSGD_hook
                 )
 
+        @unittest.skipIf(
+            BACKEND != "nccl" and BACKEND != "gloo",
+            "MPI backend does not support DDP communication hook on CUDA devices",
+        )
+        @unittest.skipIf(NO_MULTIPROCESSING_SPAWN, "Disabled for environments that \
```
Comment (rohan-varma) [inline torch/testing/_internal/distributed/distributed_test.py:3904]: Do we know why we need this skip decorator? All other tests seem to be able to run without it. Is there some detail such as use of different NCCL subgroups that cause issues when there is no spawn start method?

Comment (wayi1) [reply torch/testing/_internal/distributed/distributed_test.py:3904]: Sorry that I skipped this comment before. I don't remember the reason. Here the subgroup is equivalent to the global one, so it is really strange.
